src/GafferScene/ImageSampler.cpp

```diff
@@ -0,0 +1,447 @@
+		return inputObject;
+	}
+
+    PrimitiveVariable::IndexedView<V2f> uvView( uvVarNameIt->second );
+    const size_t size = uvView.size();
+	const PrimitiveVariable::Interpolation interpolation = uvVarNameIt->second.interpolation;
+
+	ConstStringVectorDataPtr inChannelNamesData = imagePlug()->channelNamesPlug()->getValue();
+	const std::vector<std::string> inChannelNames = inChannelNamesData->readable();
+
+	const Imath::V2i imageSize = imagePlug()->dataWindow().size();
+	const Imath::V2i imageMin = imagePlug()->dataWindow().min;
+
+    std::vector<std::string> channels;
+	IECore::StringAlgo::tokenize( channelNames, ' ', channels );
+
+	// TODO: Check for missing channels?
+
+    if( ( primVarName == "Cs" && channels.size() != 3 ) ||
+        ( primVarName == "N" && channels.size() != 3 ) ||
+        ( primVarName == "P" && channels.size() != 3 ) ||
+        ( primVarName == "Pref" && channels.size() != 3 ) ||
+        ( primVarName == "scale" && channels.size() != 3 ) ||
+        ( primVarName == "velocity" && channels.size() != 3 )
+        
+    )
+    {
+        throw IECore::Exception( "Primitive variable \"" + primVarName + "\" must sample three channels." );
+    }
+
+    if( primVarName == "uv" && channels.size() != 2 )
+    {
+        throw IECore::Exception( "Primitive variable \"uv\" must sample two channels." );
+    }
+
+    if( primVarName == "width" && channels.size() != 1)
+    {
+        throw IECore::Exception( "Primitive variable \"width\" must sample one channels." );
+    }
+
+    std::vector<OutputVariableFunction> outputs;
+
+    if( primVarName == "Cs" ||
+        primVarName == "N" ||
+        primVarName == "P" ||
+        primVarName == "Pref" ||
+        primVarName == "scale" ||
+        primVarName == "uv" ||
+        primVarName == "velocity" ||
+        primVarName == "width"
+    )
+    {
+        for( size_t i = 0; i < channels.size(); ++i )
+        {
+            if( auto o = addPrimitiveVariable( outputPrimitive.get(), primVarName, interpolation ) )
+            {
+                outputs.push_back( o );
+            }
+        }
+    }
+    else
+    {
+        for( size_t i = 0; i < channels.size(); ++i )
+        {
+            if( auto o = addPrimitiveVariable( outputPrimitive.get(), primVarName + "." + channels[i], interpolation ) )
+            {
+                outputs.push_back( o );
+            }
+        }
+    }
+
+    const int uvBoundsMode = uvBoundsModePlug()->getValue();
+
+    for( size_t channelIndex = 0; channelIndex < outputs.size(); ++channelIndex )
+    {
+        Sampler sampler( imagePlug(), channels[channelIndex], imagePlug()->dataWindow() );
+        const OutputVariableFunction &o = outputs[channelIndex];
+
+        auto rangeSampler = [ & ]( const tbb::blocked_range<size_t> &r )
+        {
+            Context::EditableScope editScope( context );
```

Comment:
I'm unsure of this part in particular, is this a valid way of handling context for the Sampler?